File: leetcode/36.cpp
//same versuion in neetcode150/1.arrays&hashing/8.Valid Sudoku.cpp
#include <array>
#include <unordered_set>
#include <vector>
using namespace std;

class Solution {
private:
    std::array<std::unordered_set<char>,9> ob,unob;//wheather hashset data in same
    std::array<std::array<std::unordered_set<char>,3>,3> sqare;
    bool go_set(char guy,int y,int x){
        if(ob[x].find(guy)!=ob[x].end()){
            return false;
        }
        ob[x].insert(guy);
        if(unob[y].find(guy)!=unob[y].end()){
            return false;
        }
        unob[y].insert(guy);
        if(sqare[y/3][x/3].find(guy)!=sqare[y/3][x/3].end()){
            return false;
        }
        sqare[y/3][x/3].insert(guy);
        return true;
    }
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        for(int y=0;y<9;++y){
            for(int x=0;x<9;++x){
                //std::cout<<board[y][x]<<" "<<y<<"   "<<x<<"\n";
                if(board[y][x]!='.' && not go_set(board[y][x],y,x)){
                    return false;
                }
            }
        }
        return true;
    }
};

File: leetcode/36.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        row=[set() for _ in range(9)]
        col=[set() for _ in range(9)]
        box=[[set() for _ in range(3)] for _ in range(3)]
        #print(row)
        #print(col)
        #print(box)
        
        #print(id(row[0]))
        #print(id(row[1]))
        #print(id(row[0])==id(row[1]))
        #return False
        index_r=-1
        for r in board:
            index_r+=1
            index_c=-1
            for c in r:
                index_c+=1
                print(index_r,index_c,c)
                if c=='.':
                    continue
                if not(c in row[index_r] or c in col[index_c] or c in box[index_r//3][index_c//3]):
                    row[index_r].add(c)
                    col[index_c].add(c)
                    box[index_r//3][index_c//3].add(c)
                else:
                    print(c)
                    print(c in row[index_r])
                    print(c in col[index_c])
                    print(c in box[index_r//3][index_c//3])
                    print(row)
                    print(col)
                    print(box)
                    print("bad")
                    return False
        return True